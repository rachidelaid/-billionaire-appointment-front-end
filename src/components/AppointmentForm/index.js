/* eslint-disable jsx-a11y/click-events-have-key-events */
import { useNavigate } from 'react-router-dom';
import { useState } from 'react';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import capitalize from './helper';
import style from './style.module.css';

/* eslint-disable camelcase */
const AppointmentForm = ({ props }) => {
  const [city, setCity] = useState('');
  const [date, setDate] = useState('');
  const [openSelect, setOpenSelect] = useState(false);

  const [billionaire_id, setBillionaire_id] = useState(props.currentBillionaireId || '');
  const [result, setResult] = useState({
    response: null,
    data: null,
  });

  const navigate = useNavigate();

  const createAppointment = async (e) => {
    e.preventDefault();

    let status = null;

    const appointment = {
      city, date, billionaire_id, user_id: props.currentUser.id,
    };

    await fetch('http://localhost:3000/api/appointments', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${props.currentUser.access_token}`,
      },
      body: JSON.stringify({ appointment }),
    })
      .then((resp) => {
        status = resp.ok;
        if (status) {
          navigate('/');
        }
        return resp.json();
      })
      .then((data) => {
        setResult({
          response: status,
          data,
        });
        toast.success('Appointment created successfully!');
        return data;
      });
  };

  const selectBillionaire = (id) => {
    setBillionaire_id(id);
    setOpenSelect(false);
  };

  return (
    <>
      {result.data && (
        <div className={style.errors}>
          {result.data && Object.entries(result.data).map((error) => (
            <p key={error[0]}>
              {capitalize(error[0])}
              {' '}
              {error[1]}
            </p>
          ))}
        </div>
      )}
      <form className={style.form} onSubmit={(e) => createAppointment(e)}>
        <input
          className={style['form-child']}
          type="text"
          placeholder="City"
          value={city}
          onChange={(e) => setCity(e.target.value)}
          required
        />
        <div className={style.select}>
          {openSelect && (
            <div className={style['select-options']}>
              {props.billionaires.map(({ id, name }) => (
                <span
                  key={id}
                  value={id}
                  role="presentation"
                  onClick={() => selectBillionaire(id)}
                >
                  {name}
                </span>
              ))}
            </div>
          )}
          <p className={style['select-title']} role="presentation" onClick={() => setOpenSelect(!openSelect)}>
            {billionaire_id ? props.billionaires.find((b) => b.id === billionaire_id).name : 'Select a billionaire'}
          </p>
        </div>
        <input
          className={style['form-child']}
          type="date"
          value={date}
          onChange={(e) => setDate(e.target.value)}
          required
          placeholder="Select date"
        />
        <button
          className={style.btn}
          type="submit"
        >
          Book Now
        </button>
      </form>
    </>
  );
};
/* eslint-enable camelcase */

AppointmentForm.defaultProps = {
  currentBillionaireId: null,
  currentUser: null,
  billionaires: [],
};

AppointmentForm.propTypes = {
  props: PropTypes.instanceOf(Object).isRequired,
  billionaires: PropTypes.arrayOf(PropTypes.shape(PropTypes.object)), // Later will be .isRequired
  currentBillionaireId: PropTypes.number,
  currentUser: PropTypes.instanceOf(Object),
};

export default AppointmentForm;
