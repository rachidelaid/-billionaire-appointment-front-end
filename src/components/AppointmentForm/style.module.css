.form {
  color: #fff;
  margin-top: 1rem;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  gap: 1.5rem;
}

.form-child,
.btn {
  border: 1px solid #fff;
  border-radius: 2rem;
  padding: 1rem;
  text-align: center;
}

.form-child {
  background-color: rgba(151, 191, 15, 0.1);
}

.btn {
  background-color: #fff;
  color: var(--green);
  grid-column: 1 / span 3;
  width: 200px;
  margin: auto;
}

::placeholder {
  color: #fff;
  opacity: 1;
}

::-ms-input-placeholder {
  color: #fff;
}

:-ms-input-placeholder {
  color: #fff;
}

.errors {
  display: flex;
  flex-direction: column;
  align-items: center;
  color: #fff;
}

@media only screen and (max-width: 670px) {
  .form {
    grid-template-columns: 1fr;
  }

  .btn {
    grid-column: unset;
    width: unset;
    margin: unset;
  }
}
