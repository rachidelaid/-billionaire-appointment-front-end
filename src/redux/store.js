import { configureStore } from '@reduxjs/toolkit';
import userReducer from './users';
import billionaireReducer from './billionaires';

const store = configureStore({
  reducer: {
    billionaires: billionaireReducer,
    users: userReducer,
  },
});

export default store;
