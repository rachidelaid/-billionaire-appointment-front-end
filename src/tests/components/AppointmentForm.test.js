import {
  fireEvent, render, screen,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import AppointmentForm from '../../components/AppointmentForm';
import store from '../../redux/store';

const props = {
  billionaires: [{
    id: 3,
    name: 'Ariel Camus',
  }],
  currentBillionaireId: 3,
  currentUser: {
    id: 1,
    access_token: 'x29mjjSGEQ88PsDCsUwKOGvB3XVsgQ8CYMe4vQA6rjs',
  },
};

const Wrapper = () => (
  <Provider store={store}>
    <BrowserRouter>
      <AppointmentForm props={props} />
    </BrowserRouter>
  </Provider>
);

describe('Appointments Component', () => {
  it('should render select tag', () => {
    render(<Wrapper />);
    const billionaireSelect = screen.getByRole('presentation');

    expect(billionaireSelect).toMatchSnapshot();
  });

  it('should render input date', () => {
    render(<Wrapper />);
    const dateInput = screen.getByPlaceholderText(/select date/i);

    expect(dateInput).toMatchSnapshot();
    fireEvent.change(dateInput, { target: { value: '2022-05-05' } });

    expect(dateInput.value).toBe('2022-05-05');
  });

  it('should render submit button', () => {
    render(<Wrapper />);
    const submitButton = screen.getByRole('button', { name: /book now/i });

    expect(submitButton).toMatchSnapshot();
  });

  it('should not create an appointment', () => {
    render(<Wrapper />);
    const submitButton = screen.getByRole('button', { name: /book now/i });

    fireEvent.click(submitButton);

    expect(submitButton).toMatchSnapshot();
  });

  it('should create an appointment', async () => {
    render(<Wrapper />);

    const submitButton = screen.getByRole('button', { name: /book now/i });

    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: 'London' },
    });

    fireEvent.click(screen.getByRole('presentation'));
    const options = await screen.findAllByText('Ariel Camus');
    fireEvent.click(options[0]);

    fireEvent.change(screen.getByPlaceholderText(/select date/i), { target: { value: '2022-05-05' } });

    fireEvent.submit(submitButton);

    expect(screen.findByRole('heading', { name: /navbar/i })).toMatchSnapshot(); // redirects to home page
  });
});
