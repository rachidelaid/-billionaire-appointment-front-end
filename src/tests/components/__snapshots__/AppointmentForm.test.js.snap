// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appointments Component should create an appointment 1`] = `Promise {}`;

exports[`Appointments Component should not create an appointment 1`] = `
<button
  class="btn"
  type="submit"
>
  Book Now
</button>
`;

exports[`Appointments Component should render input date 1`] = `
<input
  class="form-child"
  placeholder="Select date"
  required=""
  type="date"
  value=""
/>
`;

exports[`Appointments Component should render select tag 1`] = `
<p
  class="select-title"
  role="presentation"
>
  Ariel Camus
</p>
`;

exports[`Appointments Component should render submit button 1`] = `
<button
  class="btn"
  type="submit"
>
  Book Now
</button>
`;
